File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.py
import heapq

def solution(book_time):
    # "HH:MM" → 분으로 변환하는 함수
    def to_minutes(t):
        h, m = map(int, t.split(":"))
        return h * 60 + m

    # 1️⃣ 각 예약을 [시작, 종료+10] 형태로 변환
    times = []
    for start, end in book_time:
        times.append([to_minutes(start), to_minutes(end) + 10])

    # 2️⃣ 시작 시각 기준 정렬
    times.sort(key=lambda x: x[0])

    # 3️⃣ 최소 힙 (방의 청소 완료 시각)
    rooms = []

    for start, end in times:
        # 가장 빨리 끝나는 방이 현재 예약 시작 이전에 끝났으면 재사용
        if rooms and rooms[0] <= start:
            heapq.heappop(rooms)
        heapq.heappush(rooms, end)

    # 4️⃣ 힙의 크기 = 필요한 최소 객실 수
    return len(rooms)

File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.swift
import Foundation

func solution(_ book_time: [[String]]) -> Int {
    func toMin(_ t: String) -> Int {
        let parts = t.split(separator: ":")
        return Int(parts[0])! * 60 + Int(parts[1])!
    }

    // (start, end+10)로 변환 후 start 오름차순 정렬
    var books: [(Int, Int)] = book_time.map {
        let s = toMin($0[0])
        let e = toMin($0[1]) + 10
        return (s, e)
    }
    books.sort { $0.0 < $1.0 }

    // 최소 힙 구현 (끝나는 시간들)
    struct MinHeap {
        var a: [Int] = []
        var isEmpty: Bool { a.isEmpty }
        var peek: Int { a[0] }

        mutating func push(_ x: Int) {
            a.append(x)
            var i = a.count - 1
            while i > 0 {
                let p = (i - 1) / 2
                if a[p] <= a[i] { break }
                a.swapAt(p, i)
                i = p
            }
        }

        mutating func pop() -> Int {
            let res = a[0]
            let last = a.removeLast()
            if !a.isEmpty {
                a[0] = last
                var i = 0
                while true {
                    let l = i * 2 + 1
                    let r = l + 1
                    var smallest = i
                    if l < a.count && a[l] < a[smallest] { smallest = l }
                    if r < a.count && a[r] < a[smallest] { smallest = r }
                    if smallest == i { break }
                    a.swapAt(i, smallest)
                    i = smallest
                }
            }
            return res
        }
    }

    var heap = MinHeap()
    var answer = 0

    for (start, end) in books {
        // 지금 시작(start) 전에 끝난 방들은 재사용 가능
        while !heap.isEmpty && heap.peek <= start {
            _ = heap.pop()
        }
        heap.push(end)
        answer = max(answer, heap.a.count)
    }

    return answer
}